### PR TITLE
[Azure Search] Add EntityRecognition skill and ability to attach cognitive services to skillset

### PR DESCRIPTION
**Pull request by arv100kri**
1. Add `EntityRecognitionSkill` which will be the supported skill going forward in the place of `NamedEntityRecognitionSkill`
2. Add the ability to attach a cognitive services resource to a skillset
3. Fix up incorrect formatting

</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by arv100kri** [comment]
FYI: @brjohnstmsft @mgottein

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 1]
```diff
@@ -566,49 +566,16 @@
         }
       }
     },
-	"/skillsets('{skillsetName}')": {
-      "get": {
```
Comment (brjohnstmsft) [inline specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: Please re-order content in separate commits to make it easier to diff.

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 2]
```diff
@@ -566,49 +566,16 @@
...
-        "tags": ["Skillsets"],
+        "tags": [
+          "Skillsets"
+        ],
         "operationId": "Skillsets_CreateOrUpdate",
         "x-ms-examples": {
           "SearchServiceCreateOrUpdateSkillset": { "$ref": "./examples/SearchServiceCreateOrUpdateSkillset.json" }
         },
-        "description": "Creates a new cognitive skillset in an Azure Search service.",
+        "description": "Creates a new Azure search skillset or updates a skillset if it already exists.",
```
Comment (brjohnstmsft) [inline specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: @arv100kri There was a fair amount of back-and-forth about the branding of Cognitive Search and how it should be reflected here. Please make sure Luis is ok with this change.

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 3]
```diff
@@ -713,20 +721,20 @@
         "x-ms-request-id": "request-id",
         "responses": {
           "200": {
-            "description": "The list is successfully returned.",
+            "description": "he list is successfully returned.",
```
Comment (brjohnstmsft) [inline specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: Typo

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 4]
```diff
@@ -713,20 +721,20 @@
...
             "schema": {
               "$ref": "#/definitions/SkillsetListResult"
             }
           }
-		}
+        }
```
Comment (brjohnstmsft) [inline specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: Please keep formatting changes in separate commits

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 5]
```diff
@@ -4168,486 +4176,590 @@
...
-          "name": "ImageDetail",
-          "modelAsString": false
-        },
-		"description": "A string indicating which domain-specific details to return."
-	},
-	"NamedEntityCategory": {
-		"type": "string",
-        "enum": [
-          "location",
-          "organization",
```
Comment (brjohnstmsft) [inline specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: It's very difficult to tell what changed here. Please separate formatting changes into a separate commit.

**Review by bitsforthought — CHANGES_REQUESTED** [review]
@arv100kri As it stands, the mix of formatting and content changes makes this too difficult to effectively review. Please separate formatting changes into a separate commit.

**Comment by brjohnstmsft** [comment]
@arv100kri Regarding the validation failure, you can see the error message in the build output:

```
code: OBJECT_MISSING_REQUIRED_PROPERTY_DEFINITION
message: 'Missing required property definition: type'
path:
  - definitions
  - CognitiveServicesByKey
jsonPosition:
  line: 4238
  column: 31
directives: {}
jsonUrl: >-
  specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json
```

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json (hunk 2 above)
Comment (arv100kri) [reply specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: I wasn't aware of that... let me rephrase keeping the original intent intact

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json (hunk 1 above)
Comment (arv100kri) [reply specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: Do take a look at the new commits.. they should hopefully be more self-evident

**Review by bitsforthought — CHANGES_REQUESTED** [review]

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 6]
```diff
@@ -4391,6 +4436,41 @@
...
+        "minimumPrecision": {
+          "type": "number",
+          "format": "double",
+          "x-nullable": true,
+          "description": "Unused at the moment. Reserved for future use."
+        },
+        "includeTypelessEntities": {
+          "type": "boolean",
+          "x-nullable": true,
+          "description": "Determines whether or not to include entities which are well known but don't conform to a type"
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: This reflects the same choice we made in the service
I can remove this, shouldn't be an issue -

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 7]
```diff
@@ -4391,6 +4436,41 @@
...
+        },
+        "defaultLanguageCode": {
+          "$ref": "#/definitions/EntityRecognitionSkillLanguage",
+          "description": "A value indicating which language code to use. Default is en."
+        },
+        "minimumPrecision": {
+          "type": "number",
+          "format": "double",
+          "x-nullable": true,
+          "description": "Unused at the moment. Reserved for future use."
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: Maybe you can provide guidance here... the reason for including this here is to provide the most seamless transition from NamedEntityRecognitionSkill (NER) to EntityRecognitionSkill (entity)...

NERSkill filters results based on this, but the entity skill doesn't *at the moment*
This could change in the future, when that functionality is available and therefore is part of the spec... I'd like to leave it as is, but let me know if the reasoning is not sound

[the remark this answers is not in the document]
Comment (arv100kri) [reply specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: Of course I'll update the description

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json [hunk 8]
```diff
@@ -4391,6 +4436,41 @@
         },
       "description": "A skill for merging two or more strings into a single unified string, with an optional user-defined delimiter separating each component part."
     },
+    "EntityRecognitionSkill": {
+      "x-ms-discriminator-value": "#Microsoft.Skills.Text.EntityRecognitionSkill",
+      "allOf": [{
```
[the remark this answers is not in the document]
Comment (arv100kri) [reply specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: I'll correct all offenses

**Review by bitsforthought — CHANGES_REQUESTED** [review]
Still a few outstanding issues re: why a certain property is nullable, and removing a "reserved for future use" property.

File: specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (arv100kri) [reply specification/search/data-plane/Microsoft.Azure.Search.Service/preview/2017-11-11-preview/searchservice.json]: Again, the property was added to help ease transition from NERSkill to the Entity skill... The user is only required at most times to change the @odata.type property...

However, I do see the merit of your point in the swagger spec/sdk generation path.
With the compilers assistance, when a user migrates from nerskill to entity skill, it should be easy enough for them remove the parameter without having to make an API. So I'll remove it

---
In reply to: [254071834](https://github.com/Azure/azure-rest-api-specs/pull/5145#discussion_r254071834) [](ancestors = 254071834)

**Comment by arv100kri** [comment]
> Still a few outstanding issues re: why a certain property is nullable, and removing a "reserved for future use" property.

I think I've incorporated that feedback. Let me know otherwise...

**Review by bitsforthought — APPROVED** [review]

**Comment by brjohnstmsft** [comment]
@jianghaolu We're ready for you to review this now.

**Review by jianghaolu — APPROVED** [review]
